server/monitor.go

```diff
@@ -0,0 +1,504 @@
+		}
+		carr = carr[minoff:maxoff]
+
+		for _, c := range carr {
+			cli := clients[c.ID]
+			c.HBInbox = cli.HbInbox
+			if subsOption == 1 {
+				srvCli := cli.UserData.(*client)
+				c.Subscriptions = getMonitorClientSubs(srvCli, true)
+			}
+		}
+		clientsz := &Clientsz{
+			ClusterID: s.info.ClusterID,
+			ServerID:  s.serverID,
+			Now:       time.Now(),
+			Offset:    offset,
+			Limit:     limit,
+			Total:     totalClients,
+			Count:     len(carr),
+			Clients:   carr,
+		}
+		sendResponse(w, r, clientsz)
+	}
+}
+
+func getMonitorClient(c *stores.Client, subsOption int) *Clientz {
+	cli := c.UserData.(*client)
+	cli.RLock()
+	if cli.unregistered {
+		cli.RUnlock()
+		return nil
+	}
+	cz := &Clientz{
+		HBInbox: c.HbInbox,
+		ID:      c.ID,
+	}
+	if subsOption == 1 {
+		cz.Subscriptions = getMonitorClientSubs(cli, false)
+	}
+	cli.RUnlock()
+	return cz
+}
+
+func getMonitorClientSubs(client *client, needsLock bool) map[string][]*Subscriptionz {
+	if needsLock {
+		client.RLock()
+		defer client.RUnlock()
+		if client.unregistered {
+			return nil
+		}
+	}
+	subs := client.subs
+	var subsz map[string][]*Subscriptionz
+	for _, sub := range subs {
+		if subsz == nil {
+			subsz = make(map[string][]*Subscriptionz)
+		}
+		array := subsz[sub.subject]
+		newArray := append(array, createSubscriptionz(sub))
+		if &newArray != &array {
+			subsz[sub.subject] = newArray
+		}
+	}
+	return subsz
+}
+
+func getMonitorChannelSubs(ss *subStore) []*Subscriptionz {
+	ss.RLock()
+	defer ss.RUnlock()
+	subsz := make([]*Subscriptionz, 0)
+	for _, sub := range ss.psubs {
+		subsz = append(subsz, createSubscriptionz(sub))
+	}
+	for _, sub := range ss.durables {
+		subsz = append(subsz, createSubscriptionz(sub))
+	}
+	for _, qsub := range ss.qsubs {
+		qsub.RLock()
+		for _, sub := range qsub.subs {
+			subsz = append(subsz, createSubscriptionz(sub))
+		}
+		qsub.RUnlock()
+	}
+	return subsz
+}
+
+func createSubscriptionz(sub *subState) *Subscriptionz {
+	sub.RLock()
+	subz := &Subscriptionz{
+		Inbox:        sub.Inbox,
+		AckInbox:     sub.AckInbox,
+		DurableName:  sub.DurableName,
+		QueueName:    sub.QGroup,
+		IsDurable:    sub.IsDurable,
+		MaxInflight:  int(sub.MaxInFlight),
+		AckWait:      int(sub.AckWaitInSecs),
+		LastSent:     sub.LastSent,
+		PendingCount: len(sub.acksPending),
+		IsStalled:    sub.stalled,
+	}
+	sub.RUnlock()
+	return subz
+}
+
+// When we support only Go 1.8+, replace sort with sort.Slice
+type byName []string
+
+func (a byName) Len() int           { return (len(a)) }
+func (a byName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byName) Less(i, j int) bool { return a[i] < a[j] }
+
+type byChannelName []*Channelz
+
+func (a byChannelName) Len() int           { return (len(a)) }
+func (a byChannelName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byChannelName) Less(i, j int) bool { return a[i].Name < a[j].Name }
+
+func (s *StanServer) handleChannelsz(w http.ResponseWriter, r *http.Request) {
```

Comment:
This function is fairly large with lots of branching and could use some decomposing for readability/maintainability purposes IMO.